Kernel/Storage/SD/SDHostController.cpp

```diff
@@ -0,0 +1,667 @@
+{
+    // SDHC 3.7.1.2 The Sequence to Finalize a Command
+
+    // 1. Wait for the Command Complete Interrupt. If the Command Complete
+    // Interrupt has occurred,
+    //    go to step (2).
+    if (!retry_with_timeout(
+            [&]() { return m_registers->interrupt_status & command_complete; })) {
+        return EIO;
+    }
+
+    // 2. Write 1 to Command Complete in the Normal Interrupt Status register to
+    // clear this bit
+    m_registers->interrupt_status = command_complete;
+
+    // 3. Read the Response register(s) to get the response.
+    Response r = {};
+    auto cmd = last_sent_command();
+    switch (cmd.response_type) {
+    case SD::ResponseType::NoResponse:
+        break;
+    case SD::ResponseType::ResponseOf136Bits:
+        r.response[0] = m_registers->response_0;
+        r.response[1] = m_registers->response_1;
+        r.response[2] = m_registers->response_2;
+        r.response[3] = m_registers->response_3;
+        break;
+    case SD::ResponseType::ResponseOf48Bits:
+        r.response[0] = m_registers->response_0;
+        break;
+    case SD::ResponseType::ResponseOf48BitsWithBusy:
+        // FIXME: Idk what to do here
+        break;
+    }
+
+    // 4. Judge whether the command uses the Transfer Complete Interrupt or not.
+    // If it uses Transfer
+    //    Complete, go to step (5). If not, go to step (7).
+    if (last_sent_command().uses_transfer_complete_interrupt()) {
+        // 5. Wait for the Transfer Complete Interrupt. If the Transfer Complete
+        // Interrupt has occurred, go to step (6).
+
+        while ((m_registers->interrupt_status & transfer_complete) == 0)
+            ;
+
+        // 6. Write 1 to Transfer Complete in the Normal Interrupt Status
+        // register to clear this bit.
+        m_registers->interrupt_status = transfer_complete;
+    }
+
+    if (cmd.response_type != SD::ResponseType::ResponseOf136Bits) {
+        if (card_status_contains_errors(cmd.index, r.response[0])) {
+            return EIO;
+        }
+    }
+
+    // NOTE: Steps 7, 8 and 9 consist of checking the response for errors, which
+    // are specific to each command therefore those steps are not implemented
+    // here.
+
+    return { r };
+}
+
+ErrorOr<void> SDHostController::sd_clock_supply(u64 frequency)
+{
+    // SDHC 3.2.1 SD Clock Supply Sequence
+    // The *Clock Control* register is in the lower 16 bits of *Host
+    // Configuration 1*
+    VERIFY((m_registers->host_configuration_1 & sd_clock_enable) == 0);
+
+    // 1. Find out the divisor to determine the SD Clock Frequency
+    const u32 sd_clock_frequency = TRY(retrieve_sd_clock_frequency());
+
+    // FIXME: The way the SD Clock is to be calculated is different for other
+    // versions
+    VERIFY(host_version() == SD::HostVersion::Version3);
+    u32 divisor = AK::max(sd_clock_frequency / (frequency), 2);
+
+    if (sd_clock_frequency / divisor >= frequency) {
+        divisor++;
+    }
```

Comment:
could this be achieved with a `ceil_div`?

Reply:
(If I read that algorithm correctly)

Reply:
Yes